crates/wasmi/src/store.rs

```diff
@@ -714,6 +758,67 @@ impl<T> Store<T> {
         self.data
     }
 
+    pub fn limiter(
+        &mut self,
+        mut limiter: impl FnMut(&mut T) -> &mut (dyn crate::ResourceLimiter) + Send + Sync + 'static,
+    ) {
+        let inner = &mut self.inner;
+        let (instance_limit, table_limit, memory_limit) = {
+            let l = limiter(&mut self.data);
+            (l.instances(), l.tables(), l.memories())
+        };
+        inner.instance_limit = instance_limit;
+        inner.table_limit = table_limit;
+        inner.memory_limit = memory_limit;
+        self.limiter = Some(ResourceLimiterQuery(Box::new(limiter)))
+    }
+
+    pub fn bump_resource_counts(&mut self, module: &crate::Module) -> Result<(), Error> {
```

Comment:
Instead of immediately bumping the counters we could make this method just check if creating a new instance for a Wasm module fits the limits as early check and potential bail out during the instantiation process.

This way we don't have the need for those redundant counters but also get our cheap checks. Or am I missing something?

Reply:
No, I think that's possible. I can give it a try, I just was trying to follow how wasmtime did it / when it did the checks.

Reply:
low priority issue though. highest priorities are getting CI happy and fixing those benchmarks. everything else can be done later.

Reply:
Update: this does seem to work ok if I move some stuff around, I've pushed a version that does so.